cl/validator/attestation_producer/attestation_producer.go

```diff
@@ -38,18 +41,110 @@ type attestationProducer struct {
 	if err != nil {
 		panic(err)
 	}
 
-	return &attestationProducer{
+	p := &attestationProducer{
 		beaconCfg:         beaconCfg,
 		attestationsCache: attestationsCache,
 	}
+	go p.listenOnCopyStateAtSlot(ctx)
+	return p
+}
+
+// listenOnCopyStateAtSlot generates beacon state at slot in background and also notify all ch in waiting list. This is used to avoid over-copying of
+// beacon state in case of multiple requests for the same slot. (e.g. 10k validators requesting attestation data for the same slot almost at the same time,
+// which might lead to 10k copies of the same state in memory)
+func (ap *attestationProducer) listenOnCopyStateAtSlot(ctx context.Context) {
+	// beaconStateCache *lru.CacheWithTTL[uint64, *state.CachingBeaconState] // Slot => BeaconState
+	beaconStateCache := lru.NewWithTTL[uint64, *state.CachingBeaconState]("attestaion_producer_beacon_state_cpy", 32, time.Minute)
+	generatingWaitList := map[uint64][]chan<- stateAcquireResp{}
+	slotGenerateDone := make(chan uint64, 32)
+	for {
+		select {
+		case req := <-ap.reqStateAtSlotCh:
+			slot := req.slot
+			baseState := req.baseState
+			if state, ok := beaconStateCache.Get(slot); ok {
+				// if state is already generated, return it
+				req.resp <- stateAcquireResp{state: state, err: nil}
+				continue
+			}
+			if _, ok := generatingWaitList[slot]; ok {
+				// append to wait list
+				generatingWaitList[slot] = append(generatingWaitList[slot], req.resp)
+				continue
+			}
+			generatingWaitList[slot] = []chan<- stateAcquireResp{req.resp}
+
+			// generate state in background
+			go func() {
+				defer func() {
+					// notify all waiting requests
+					slotGenerateDone <- slot
+				}()
+				cpyBaseState, err := baseState.Copy()
+				if err != nil {
+					log.Warn("Failed to copy base state", "slot", slot, "err", err)
+					return
+				}
+				if err := transition.DefaultMachine.ProcessSlots(cpyBaseState, slot); err != nil {
+					log.Warn("Failed to process slots", "slot", slot, "err", err)
+					return
+				}
+				beaconStateCache.Add(slot, cpyBaseState)
+			}()
+		case slot := <-slotGenerateDone:
+			if reqChs, ok := generatingWaitList[slot]; ok {
+				delete(generatingWaitList, slot)
+				state, ok := beaconStateCache.Get(slot)
+				if !ok {
+					log.Warn("Failed to get generated state", "slot", slot)
+					for _, ch := range reqChs {
+						ch <- stateAcquireResp{state: nil, err: errors.New("failed to get generated state")}
+					}
+					continue
+				}
+				for _, ch := range reqChs {
+					ch <- stateAcquireResp{state: state, err: nil}
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// acquireBeaconStateAtSlot returns a copy of the beacon state at the given slot. Note the returned state should NOT be modified.
+func (ap *attestationProducer) acquireBeaconStateAtSlot(baseState *state.CachingBeaconState, slot uint64) (*state.CachingBeaconState, error) {
```

Comment:
I see... this looks a bit over complicated though, I want to suggest a different approach:

This is the problematic bit:
```
if stateEpoch < epoch {
		baseState, err = baseState.Copy()
		if err != nil {
			return solid.AttestationData{}, err
		}
		if err := transition.DefaultMachine.ProcessSlots(baseState, slot); err != nil {
			return solid.AttestationData{}, err
		}
	}
	
```

What I would do is changing: 
1) `transition.DefaultMachine.ProcessSlots(baseState, slot)` -> `transition.DefaultMachine.ProcessSlots(baseState, stateEpoch*SlotsPerEpoch)` as the final state is going to be as good
2) Store this state into a cache with low TTL (say 4)